web/src/pill_typeahead.ts

```diff
@@ -0,0 +1,220 @@
+import assert from "minimalistic-assert";
+
+import * as blueslip from "./blueslip";
+import * as people from "./people";
+import type {User} from "./people";
+import * as stream_pill from "./stream_pill";
+import type {StreamSubscription} from "./sub_store";
+import * as typeahead_helper from "./typeahead_helper";
+import type {UserOrMention} from "./typeahead_helper";
+import * as user_group_pill from "./user_group_pill";
+import * as user_groups from "./user_groups";
+import type {UserGroup} from "./user_groups";
+import * as user_pill from "./user_pill";
+
+function person_matcher(query: string, item: User): boolean {
+    if (people.is_known_user(item)) {
+        return typeahead_helper.query_matches_person(query, item);
+    }
+    return false;
+}
+
+function group_matcher(query: string, item: UserGroup): boolean {
+    if (user_groups.is_user_group(item)) {
+        return typeahead_helper.query_matches_name(query, item);
+    }
+    return false;
+}
+
+type StreamRecipient = StreamSubscription & {type: "stream"};
+export type UserGroupRecipient = UserGroup & {type: "user_group"};
+export type UserRecipient = UserOrMention & {type: "user"};
+
+export type TypeaheadItem = StreamRecipient | UserGroupRecipient | UserRecipient;
+
+export function set_up(
+    $input: JQuery<HTMLInputElement>,
+    pills: any,
```

Comment:
This can be `StreamPillWidget` or `UserGroupPillWidget` or `UserPillWidget`, i.e. it's expecting a pill widget but not a specific one.

I see two instances of this being called with `UserPillWidget` and only being used for user pills, and a third of it being called with `UserGroupPillWidget` but setting up pills for all three kinds. So I'm a bit confused about how to settle up the types here